utilities/write_batch_with_index/write_batch_with_index.cc

```diff
@@ -819,7 +827,12 @@ Status WriteBatchWithIndex::GetFromBatchAndDB(DB* db,
          result == WriteBatchWithIndexInternal::Result::kNotFound);
 
   // Did not find key in batch OR could not resolve Merges.  Try DB.
-  s = db->Get(read_options, column_family, key, pinnable_val);
+  if (!callback) {
+    s = db->Get(read_options, column_family, key, pinnable_val);
+  } else {
+    s = reinterpret_cast<DBImpl*>(db)->GetImpl(read_options, column_family, key,
```

Comment:
This is a dirty hack to be able to pass the callback to DBImpl without having to change the public API of DB with the ReadCallback argument. I assumed that it is an ok hack since a few lines above at line 797 we do the same hack.

The reason that we ended up with this hack is that 
i) we needed to augment DBImpl, which is an internal class, to accept a ReadCallback argument
ii) Transaction implementation which supply ReadCallback are also internal and have access to DBImpl
ii) Transactions however are using the public class WriteBatchWithIndex to read from both the batch and DB
iii) Being public API, WriteBatchWithIndex operates with DB instead of the internal class DBImpl

Reply:
For line 797 this would work better: `reinterpret_cast<DBImpl*>(db->GetRootDB())->immutable_db_options()`. Let's update it.

However calling `reinterpret_cast<DBImpl*>(db->GetRootDB())->GetImpl()` seems tricky. It makes `WriteBatchWithIndex` not able to work with stackable db. How about we return NotSupported status if `callback != nullptr && db->GetRootDB() != db`?

Reply:
Even better if we can find a way to hide `ReadCallback`..

Reply:
`callback != nullptr && db->GetRootDB() != db` sounds good to me. We should apply it the existing hack as well. So I guess we should return NotSupported when `db->GetRootDB() != db`.

`ReadCallback` has a private constructor so it being non-hidden is still safe. It cannot be hidden since it has to go through a (private) method of the public class WriteBatchWithIndex, so it needs to defined in the public header of WriteBatchWithIndex.

Reply:
Actually let me take it back. I am adding a commit that hides `ReadCallback`.